### PR TITLE
Restart input connection after `EditableText.onSubmitted`

### LINKED ISSUE #84240: Keyboard mode doesn't reset onSubmitted
Internal issue b/190464855

<details>

 <summary>Repro code</summary>

```dart
import 'package:flutter/material.dart';
 
void main() => runApp(MyApp());
 
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final appTitle = 'My widget';
    return MaterialApp(
      title: appTitle,
      home: Scaffold(
        appBar: AppBar(
          title: Text(appTitle),
        ),
        body: MyWidget(),
      ),
    );
  }
}
 
class MyWidget extends StatefulWidget {
  @override
  MyWidgetState createState() {
    return MyWidgetState();
  }
}
 
class MyWidgetState extends State<MyWidget> {
  final contents = <String>[];
  final focusNode = FocusNode();
  final controller = TextEditingController();
 
  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
 
  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Expanded(
          child: ListView(
            children: contents
                .map((line) => Text(line, textAlign: TextAlign.center))
                .toList(),
            reverse: true,
          ),
        ),
        Padding(
          padding: EdgeInsets.symmetric(horizontal: 8, vertical: 16),
          child: TextField(
            focusNode: focusNode,
            controller: controller,
            decoration: InputDecoration(
              border: OutlineInputBorder(),
              hintText: 'Enter something and submit',
            ),
            textCapitalization: TextCapitalization.sentences,
            onSubmitted: (value) {
              setState(() => contents.insert(0, value));
              controller.clear();
              focusNode.requestFocus();
            },
          ),
        ),
      ],
    );
  }
}
```

</details>

Focus on the textfield, change the keyboard mode to digits, type digits and submit. The keyboard is still in digit mode vs native apps where the keyboard would have switched back to letters.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes https://github.com/flutter/flutter/issues/84240. 
Restart the input connection in case the developer tries to undo the focus change in their `Editable.onSubmitted` callback. The Android & iOS equivalents of `Editable.onSubmitted` both reset the keyboard type.

On iOS currently there doesn't seem to be a way to "reset" the keyboard except for changing the focus and then switching back to the text field (which creates a new `UITextInput` instance in the embedder).

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -1917,9 +1917,14 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       }
     }
 
+    final ValueChanged<String>? onSubmitted = widget.onSubmitted;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:1920]: This code creates a local variable to save an exclamation mark. Is it worth it?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -1928,6 +1933,17 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         context: ErrorDescription('while calling onSubmitted for $action'),
       ));
     }
+
+    // Create a new TextInputConnection to replace the current one. This
+    // on iOS switches to a new input view and on Android restarts the
+    // input method. In both cases this call resets the IME.
+    //
+    // We do this for fidelity reasons, in case the developer calls
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:1941]: avoid using `we` in comment, different people may have different interpretation of who `we` are

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -1928,6 +1933,17 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    // requestFocus in the onSubmitted callback to undo the focus change.
+    // See https://github.com/flutter/flutter/issues/84240.
+    if (shouldUnfocus) {
+      _restartConnectionIfNeeded();
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:1945]: is this the only use case where we want to restart the connection? what about programmatically changing the focus?

Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:1945]: Will _finalizeEditing still be called in that case?  It looks like anything that closes the connection will call it.  I'm also curious if there are any other cases that this PR wouldn't cover, can't think of any others though.

**Review by justinmc — APPROVED** [review]
LGTM 👍 , seems like the right way to fix this.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1945]: Yes currently that's the only case we the we want to restart the connection to restart the IME from the framework.

**Review by LongCatIsLooong — COMMENTED** [review]
@chunhtai @justinmc I updated the pull request a little bit to address the comments and put the `_restartConnectionIfNeeded` call in a microtask.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -2017,6 +2022,18 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    // If `shouldUnfocus` is true but in case the developer cancelled the focus
+    // change in onSubmitted by focusing this input field again, reset the
+    // soft keyboard. See https://github.com/flutter/flutter/issues/84240.
+    //
+    // `_restartConnectionIfNeeded` creates a new TextInputConnection to replace
+    // the current one. This on iOS switches to a new input view and on Android
+    // restarts the input method, and in both cases the soft keyboard will be
+    // reset.
+    if (shouldUnfocus) {
+      scheduleMicrotask(_restartConnectionIfNeeded);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:2035]: This looks scary, why we need to put this in a microtask? also do we need to debounce?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2035]: This is to check if the text field is indeed unfocused after the focus tree is updated. Yeah sounds like a good idea to schedule the restart only once.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart:2035]: will that be determine in the same frame? if so, we may want postframecall back instead

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Done.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: > will that be determine in the same frame? if so, we may want postframecall back instead


The focus tree is updated twice per frame so I suspect in some extreme cases doing that in a postframe callback will be too late.

The logic here is trying to restart the connection only if the text field's scheduled focus change (in response to some text input actions) was cancelled immediately by the developer (in which case the text field's focus change callbacks will not be invoked, and the `TextInputConnection` will stay connected if we don't restart).

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I thought the scheduleMicrotask will also be pushed at the end of the frame? or am I missing something?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: The queued microtasks are run twice (the idle phase and the midFrameMicrotasks phase) each frame if I understand it correctly: https://api.flutter.dev/flutter/scheduler/SchedulerPhase-class.html

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I see, so it will run before the next frame starts. SGTM

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**